torchgen/autoheuristic/gen_data_pad_mm.py

```diff
@@ -0,0 +1,208 @@
+# mypy: ignore-errors
+
+import argparse
+import random
+import time
+from typing import Any, Tuple
+
+import torch
+
+torch.set_default_device("cuda")
+from triton.testing import do_bench
+
+from torch._inductor.utils import fresh_inductor_cache
+
+# A100: 81920MiB
+# without a threshold we sometimes run out of memory
+threshold_memory = 85899345920 / 4
+
+# probability that a dimension is unaligned
+p_unaligned = 0.25
+
+# probability that a tensor is "prepadded", i.e. pad_mm excludes time it takes to pad from benchmarking
+p_prepadded = 0.2
+
+# probability that we pick from uniform distribution
+p_uniform = 0.5
+
+# probability that a tensor is transposed
+p_transposed = 0.1
+
+p_float32_prec_highest = 0.8
+
+
+def benchmark(
+    m: int,
+    k: int,
+    n: int,
+    tranpose_left: bool,
+    tranpose_right: bool,
+    dtype: Any,
+    prepadded_left: bool,
+    prepadded_right: bool,
+) -> Any:
+    if tranpose_left:
+        a = torch.randn(k, m, dtype=dtype).t()
+    else:
+        a = torch.randn(m, k, dtype=dtype)
+    if tranpose_right:
+        b = torch.randn(n, k, dtype=dtype).t()
+    else:
+        b = torch.randn(k, n, dtype=dtype)
+
+    with fresh_inductor_cache():
+
+        def mm(a, b):
+            return torch.mm(a, b)
+
+        def mm_mat1_prepadded(a, b):
+            return torch.mm(a + 1, b)
```

Comment:
Can you explain how `a + 1` achieve the prepad effect?

Reply:
I cannot explain it in detail, but I looked at how it is tested in here https://github.com/pytorch/pytorch/blob/main/test/inductor/test_pad_mm.py#L413. It makes this function return True https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/pad_mm.py#L300.

Reply:
oh, maybe the main purpose is to make the matmul input not a placeholder.